### PR TITLE
SPIN-775: Define the initial models/schema and request/response structure

### PR DESCRIPTION
**Pull request by fishnix**
Initial implementation for the *tweaser* schema and models.  All endpoints except for the actual response POST are authenticated with the PSK.  The POST endpoint is authenticated with a token generated from the QuestionID, the userid and a different PSK.

### CODE REVIEW

File: grifts/db.go [hunk 1]
```diff
@@ -1,15 +1,203 @@
...
+		return err
+	}
+
+	if _, err = newAnswer("Camden", fdq.ID); err != nil {
+		return err
+	}
+	if _, err = newAnswer("Tenyo", fdq.ID); err != nil {
+		return err
+	}
+	if _, err = newAnswer("Galen", fdq.ID); err != nil {
```
Comment (galenjdove) [inline grifts/db.go]: The obvious answer

**Comment by galenjdove** [comment]
👍

File: actions/app.go [hunk 2]
```diff
@@ -31,7 +36,7 @@ func App() *buffalo.App {
 			SessionName: "_tweaser_session",
 		})
 		// Automatically redirect to SSL
-		app.Use(forceSSL())
+		// app.Use(forceSSL())
```
Comment (tenyo) [inline actions/app.go]: ✂️

File: actions/responses.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+		// Validate the posted data and save it to the database
+		ve, err := tx.ValidateAndCreate(ra)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		verrs.Append(ve)
+	}
+
+	if verrs.HasAny() {
```
Comment (tenyo) [inline actions/responses.go]: What does this check exactly?

File: actions/responses.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		return c.Render(422, r.JSON(verrs))
+	}
+
+	log.Println("done with update")
+
+	return c.Render(202, r.JSON("submitted"))
```
Comment (tenyo) [inline actions/responses.go]: Shouldn't this be just `200`? It's not async.

File: helpers/tokens.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+// ModelToken is the object used to generate a token loosesly associated with a model
+type ModelToken struct {
+	UserID string    `json:"user_id"`
+	ID     uuid.UUID `json:"id"`
+	Secret string    `json:"secret"`
+}
+
+// Generate creates a base64 encoded renewal token
```
Comment (tenyo) [inline helpers/tokens.go]: 👀 renewal

File: helpers/tokens.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	token, err := bcrypt.GenerateFromPassword(str, bcrypt.DefaultCost)
+	if err != nil {
+		return "", err
+	}
+	log.Debugln("Secret hash:", string(token))
+
+	return base64.StdEncoding.EncodeToString(token), nil
+}
+
+// Validate validates a base64 encoded renewal token
```
Comment (tenyo) [inline helpers/tokens.go]: 👀

File: README.md [hunk 7]
```diff
@@ -1,35 +1,27 @@
...
-If you point your browser to [http://127.0.0.1:3000](http://127.0.0.1:3000) you should see a "Welcome to Buffalo!" page.
-
-**Congratulations!** You now have your Buffalo application up and running.
-
-## What Next?
-
-We recommend you heading over to [http://gobuffalo.io](http://gobuffalo.io) and reviewing all of the great documentation there.
-
-Good luck!
+`Responses` are the responses to questions from users.  A response is POST'd to the endpoint for a question.  The response may have more than one `Answer` and some text if the `Answer` type is `input`.  Responses have a `many_to_many` relationship with `Answers`.
```
Comment (tenyo) [inline README.md]: A list of all the endpoints and some examples here would be awesome!

File: actions/app.go [hunk 8]
```diff
@@ -93,7 +96,7 @@ func sharedTokenAuth(next buffalo.Handler) buffalo.Handler {
 		headers, ok := c.Request().Header["X-Auth-Token"]
 		if !ok || len(headers) == 0 || headers[0] != AdminToken {
 			log.Println("Missing or bad token header for request", c.Request().URL)
-			return c.Error(403, errors.New("Forbidden!"))
+			return c.Error(403, errors.New("Forbiden!"))
```
Comment (tenyo) [inline actions/app.go]: ❓

File: actions/questions.go [hunk 9]
```diff
@@ -18,20 +20,35 @@ func QuestionsList(c buffalo.Context) error {
...
+		for _, c := range campaigns {
+			campaignIDs = append(campaignIDs, c.ID.String())
+		}
+
+		// Paginate results. Params "page" and "per_page" control pagination.
+		// Default values are "page=1" and "per_page=20".
+		q := tx.PaginateFromParams(c.Params())
+		q = q.Where("questions.enabled = true")
+		q = q.Where("questions.campaign_id IN (?)", campaignIDs...)
+		q = q.Where("id NOT in (select question_id FROM responses WHERE user_id = (?))", userid)
```
Comment (tenyo) [inline actions/questions.go]: SQL fun 😮

File: actions/questions.go [hunk 10]
```diff
@@ -18,20 +20,35 @@ func QuestionsList(c buffalo.Context) error {
...
+		err = q.Eager("Answers").All(&questions)
```
Comment (tenyo) [inline actions/questions.go]: Should this only get the `enabled` answers?

File: actions/responses.go (hunk 3 above)
Comment (fishnix) [reply actions/responses.go]: screen shot 2018-08-27 at 1 13 26 pm

you need to get the Sourcegraph plugin 😆

File: actions/responses.go (hunk 4 above)
Comment (fishnix) [reply actions/responses.go]: I think this is appropriate actually

https://httpstatuses.com/202

File: helpers/tokens.go (hunk 5 above)
Comment (fishnix) [reply helpers/tokens.go]: ❓ ❓ ❓ ❓ 😆

File: actions/app.go (hunk 8 above)
Comment (fishnix) [reply actions/app.go]: 🤷‍♂️

File: actions/responses.go (hunk 3 above)
Comment (tenyo) [reply actions/responses.go]: 😻 👍

File: actions/responses.go (hunk 4 above)
Comment (tenyo) [reply actions/responses.go]: The processing has been all done at this point, that's why I thought 200 makes more sense

**Comment by tenyo** [comment]
Looks good 👍

File: actions/questions.go [hunk 11]
```diff
@@ -67,6 +67,14 @@ func QuestionsList(c buffalo.Context) error {
 			}
 
 			questions[i].Token = token
+
+			// Get the enabled answers for the question and append them to the questions response
+			answers := []models.Answer{}
+			err = tx.Where("question_id = ?", q.ID).Where("enabled = true").All(&answers)
```
Comment (tenyo) [inline actions/questions.go]: Move this inside the if?

**Comment by tenyo** [comment]
Let's 🚢 it!

**Pull request merged**